File: Data Structures & Algorithms/house-robber/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 3:
            return max(nums)

        n = len(nums)
        dp = nums[:2]
        dp.append(dp[0] + nums[2])

        for i in range(3, n):
            num = nums[i]
            dp.append(max(dp[i-2], dp[i-3]) + num)

        return max(dp)

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int zero_idx = -1;
        int n = nums.size();

        vector<int> res(n);

        int total = 1;
        for (int i=0;i<n;++i) {
            int num = nums[i];
            if (num==0) {
                if (zero_idx >= 0) return res;
                zero_idx = i;
            } else {
                total *= num;
            }
        }
        if (zero_idx >= 0) {
            res[zero_idx] = total;
            return res;
        }

        for (int i=0;i<n;++i) {
            res[i] = total / nums[i];
        }

        return res;
    }
};
